hudi-spark-datasource/hudi-spark3-extensions_2.12/src/main/scala/org/apache/spark/sql/hudi/execution/MergeIntoHudiTable.scala

```diff
@@ -0,0 +1,332 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.hudi.execution
+
+import org.apache.hudi.DataSourceWriteOptions.{DEFAULT_PAYLOAD_OPT_VAL, PAYLOAD_CLASS_OPT_KEY, RECORDKEY_FIELD_OPT_KEY}
+import org.apache.hudi.common.model.{OverwriteNonDefaultsWithLatestAvroPayload, WriteOperationType}
+import org.apache.hudi.execution.HudiSQLUtils
+import org.apache.spark.internal.Logging
+import org.apache.spark.sql._
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.encoders.{ExpressionEncoder, RowEncoder}
+import org.apache.spark.sql.catalyst.expressions.codegen.GeneratePredicate
+import org.apache.spark.sql.catalyst.merge._
+import org.apache.spark.sql.catalyst.expressions.{And, Attribute, BasePredicate, Expression, Literal, PredicateHelper, UnsafeProjection}
+import org.apache.spark.sql.functions.{col, lit, when}
+import org.apache.spark.sql.catalyst.merge.{HudiMergeClause, HudiMergeInsertClause}
+import org.apache.spark.sql.catalyst.plans.logical.{LocalRelation, LogicalPlan, Project}
+import org.apache.spark.sql.execution.command.RunnableCommand
+import org.apache.spark.sql.internal.StaticSQLConf
+import org.apache.spark.sql.types.{StringType, StructType}
+
+import scala.collection.mutable.ArrayBuffer
+
+/**
+  * merge command, execute merge operation for hudi
+  */
+case class MergeIntoHudiTable(
+    target: LogicalPlan,
+    source: LogicalPlan,
+    joinCondition: Expression,
+    matchedClauses: Seq[HudiMergeClause],
+    noMatchedClause: Seq[HudiMergeInsertClause],
+    finalSchema: StructType,
+    trimmedSchema: StructType) extends RunnableCommand with Logging with PredicateHelper {
+
+  var tableMeta: Map[String, String] = null
+
+  lazy val isRecordKeyJoin = {
+    val recordKeyFields = tableMeta.get(RECORDKEY_FIELD_OPT_KEY).get.split(",").map(_.trim).filter(!_.isEmpty)
+    val intersect = joinCondition.references.intersect(target.outputSet).toSeq
+    if (recordKeyFields.size == 1 && intersect.size ==1 && intersect(0).name.equalsIgnoreCase(recordKeyFields(0))) {
+      true
+    } else {
+      false
+    }
+  }
+
+  override def run(sparkSession: SparkSession): Seq[Row] = {
+    tableMeta = HudiSQLUtils.getHoodiePropsFromRelation(target, sparkSession)
+    val enableHive = "hive" == sparkSession.sessionState.conf.getConf(StaticSQLConf.CATALOG_IMPLEMENTATION)
```

Comment:
here please the logic put into a method called `hiveCatalogImplementaion`?

Reply:
agree